rubric_data/views.py

```diff
@@ -82,28 +87,32 @@ def denormalize(data):
     
         # iterate through submissions under assignments
         for submission in assignment['submissions']:
-            if 'rubric_assessment' in submission.keys() and submission['workflow_state'] == 'graded':
+            if 'rubric_assessment' in submission.keys() and submission['workflow_state'] == 'graded' and submission['score'] is not None:
                 submission_id = submission['id']
                 student_id = submission['user_id']
                 student_name = students_lookup[student_id]['sortable_name']
-                section_id = sections_lookup[student_id]
+                section_tuples = sections_lookup[student_id]
 
                 # iterate through criteria and ratings under submissions
                 for criterion_id, criterion_data in submission['rubric_assessment'].items():
-                    score = criterion_data['points']
-                    criterion_name = criteria_lookup[criterion_id]['description']
-                    row = {
-                        "student_id": student_id,
-                        "student_name": student_name,
-                        "assignment_id": assignment_id,
-                        "assignment_name": assignment_name,
-                        "criterion_id": criterion_id,
-                        "criterion_name": criterion_name,
-                        "section_id": section_id,
-                        "score": score,
-                        "rating": get_rating(criterion_id, score, criteria_lookup)
-                    }
-                    output.append(row)
+                    if 'points' in criterion_data.keys() and criterion_data['rating_id'] is not None:
+                        score = criterion_data['points']
+                        unique_criterion_id = f"{assignment_id}{criterion_id}"
+                        criterion_name = criteria_lookup[unique_criterion_id]['description']
+                        for section_tuple in section_tuples:
```

Comment:
Does the visualization de-duplicate in any way, or just aggregate whatever is in the denormalized data? If it doesn't de-duplicate, and students are enrolled in different numbers of sections, then the aggregate visualization would end up weighting the data by the number of sections each student is enrolled in.

Reply:
Smart.

Reply:
It doesn't de-duplicate at all, since in the denormalized data a student should be represented across multiple sections if that is what is represented in the data from canvas. But you are right that when we aggregate all sections, we should only use a student once. But I think that should happen in the vis data pivoting that happens in JS. I'll make the change 👍